### PR TITLE
Implement support giver menu business logic

### PR DESCRIPTION
**Pull request by lidorz51**

### CODE REVIEW

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 1]
```diff
@@ -25,11 +25,38 @@ bool filterByStatus(void* ticket, void* status)
...
+bool filterByStakeholders(void* ticket1, void*ticket2)
+{
+	return wcscmp(ticket_getstakeholders(ticket1), ticket_getstakeholders(ticket2))!= 0;
+}
+bool searchByCustomerEmail(void* ticket1, void*ticket2)
+{
+	return wcscmp(ticket_getCustomerEmail(ticket1), ticket_getCustomerEmail(ticket2)) != 0;
+
+}
+bool searchBytags(void* ticket1, void*ticket2)
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Function name does not conform to code style guide.

File: Source Code/Ticket Management System/Ticket Management System.vcxproj [hunk 2]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (soryy708) [inline Source Code/Ticket Management System/Ticket Management System.vcxproj]: No change

File: Source Code/Ticket Management System/Ticket Management System.vcxproj [hunk 3]
```diff
@@ -22,7 +22,7 @@
     <VCProjectVersion>15.0</VCProjectVersion>
     <ProjectGuid>{BCF3B1AC-70F1-40FD-8EF0-6730B6E2D4E7}</ProjectGuid>
     <RootNamespace>TicketManagementSystem</RootNamespace>
-    <WindowsTargetPlatformVersion>10.0.17763.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>8.1</WindowsTargetPlatformVersion>
```
Comment (soryy708) [inline Source Code/Ticket Management System/Ticket Management System.vcxproj]: Unnecesary change.

File: Source Code/Ticket Management System/Ticket Management System.vcxproj [hunk 4]
```diff
@@ -41,7 +41,7 @@
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'" Label="Configuration">
     <ConfigurationType>Application</ConfigurationType>
     <UseDebugLibraries>true</UseDebugLibraries>
-    <PlatformToolset>v141</PlatformToolset>
+    <PlatformToolset>v140</PlatformToolset>
```
Comment (soryy708) [inline Source Code/Ticket Management System/Ticket Management System.vcxproj]: Unnecesary change.

File: Source Code/Ticket Management System/i18n.h [hunk 5]
```diff
@@ -32,6 +32,11 @@ enum i18n_StringId
 	I18N_STRING_SUPPORTGIVER_VIEWTICKET,
 	I18N_STRING_SUPPORTGIVER_BROWSETICKETS,
 	I18N_STRING_SUPPORTGIVER_CREATEUSER,
+	I18N_STRING_SUPPORTGIVER_CUSTOMEREMILE,
```
Comment (soryy708) [inline Source Code/Ticket Management System/i18n.h]: Who is Emile?

File: Source Code/Ticket Management System/i18n.h [hunk 6]
```diff
@@ -32,6 +32,11 @@ enum i18n_StringId
...
+	I18N_STRING_SUPPORTGIVER_TITLE,
+	I18N_STRING_SUPPORTGIVER_TYPE,
+	I18N_STRING_SUPPORTGIVER_SERVERITY,
+	I18N_STRING_SUPPORTGIVER_DI,
```
Comment (soryy708) [inline Source Code/Ticket Management System/i18n.h]: What's a DI?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 7]
```diff
@@ -25,11 +25,38 @@ bool filterByStatus(void* ticket, void* status)
 {
 	return wcscmp(ticket_getStatus(ticket), status) != 0;
 }
+bool filterByTier(void* ticket1, void* ticket2)
+{
+	return wcscmp(ticket_getTier(ticket1), ticket_getTier(ticket2)) != 0;
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Why filterByTier gets 2 tickets?
Why not ticket and tier?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 8]
```diff
@@ -25,11 +25,38 @@ bool filterByStatus(void* ticket, void* status)
...
+}
+bool filterByType(void* ticket1, void* ticket2)
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Why filterByType gets 2 tickets?
Why not ticket and type?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 9]
```diff
@@ -25,11 +25,38 @@ bool filterByStatus(void* ticket, void* status)
...
+bool sortBySeverity(void* ticket1, void* ticket2)
+{
+	return wcscmp(ticket_getSeverity(ticket1), ticket_getSeverity(ticket2)) > 0;
 
+}
 bool sortByTitle(void* ticket1, void* ticket2)
 {
 	return wcscmp(ticket_getTitle(ticket1), ticket_getTitle(ticket2)) > 0;
 }
+bool filterByStakeholders(void* ticket1, void*ticket2)
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Why filterByStakeholders gets 2 tickets?
Why not ticket and stakeholders?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 10]
```diff
@@ -77,30 +104,74 @@ void browseTickets()
 					}
 					case('2'): // Filter by tier
 					{
-						// TODO
+						wchar_t* cty;
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: What's a `cty`?
You mean `ctx`, which is short for `context`.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 11]
```diff
@@ -77,30 +104,74 @@ void browseTickets()
...
+						wcstok(tier, "\n", &cty);
+						struct Vector* filtered_vector = vector_filter(tickets, filterByTier, tier);
+						printTicketsArray(filtered_vector);
+						vector_destroy(filtered_vector);
 						break;
 					}
 					case('3'): // Filter by type
 					{
-						// TODO
+						wchar_t* ctz;
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: What's a `ctz`?
You mean `ctx`, which is short for `context`.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 12]
```diff
@@ -129,17 +200,54 @@ void browseTickets()
 	}
 }
 
+void Opentickets()
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Function name does not conform to code style guide.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 13]
```diff
@@ -129,17 +200,54 @@ void browseTickets()
...
+{
+		wchar_t customerEmail[256];
+		wchar_t title_support[256];
+		wchar_t type_support[256];
+		wchar_t severity[] = { "Low", "Medium", "High", "Very high" };
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: This does not compile.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 14]
```diff
@@ -129,17 +200,54 @@ void browseTickets()
...
+		wchar_t description_support[512];
+
+		system("CLS");
+		wscanf(L"%s\n", customerEmail);
+		wprintf(L"%s\n", i18n_getString(I18N_STRING_CUSTOMER_CONTACTSUPPORT, I18N_LOCALE_CURRENT));
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Support giver is not customer. The output must be different.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 15]
```diff
@@ -129,17 +200,54 @@ void browseTickets()
...
+		wprintf(L"===============\n");
+		wprintf(L"%s\n", i18n_getString(I18N_STRING_CUSTOMER_CONTACTSUPPORT_DESCRIPTION, I18N_LOCALE_CURRENT));
+		wscanf(L"%s", description_support);
+		wprintf(L"%s\n", i18n_getString(I18N_STRING_CUSTOMER_CONTACTSUPPORT_TITLE, I18N_LOCALE_CURRENT));
+		wscanf(L"%s", title_support);
+		wprintf(L"%s\n", i18n_getString(I18N_STRING_CUSTOMER_CONTACTSUPPORT_TYPE, I18N_LOCALE_CURRENT));
+		wscanf(L"%s", type_support);
+
+		struct Ticket* ticket = ticket_create(customerEmail,title_support, type_support,severity,description_support, auth_getEmail());
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: `severity` is a logic error that will, scarily, compile.

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 13 above)
Comment (soryy708) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Also, there is no input for severity.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 16]
```diff
@@ -77,30 +104,74 @@ void browseTickets()
...
+						wprintf(L"%s\n", L"please input type");
+						wscanf(L"%s", type);
+						wcstok(type, "\n", &ctz);
+						struct Vector* filtered_vector = vector_filter(tickets, filterByType, type);
+						printTicketsArray(filtered_vector);
+						vector_destroy(filtered_vector);
 						break;
 					}
-					case('4'): // Filter by stakeholders
+					case('4'): // sort by severity
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Case numbers changed. Was menu creation updated to reflect new meaning of inputs?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 17]
```diff
@@ -77,30 +104,74 @@ void browseTickets()
...
-						// TODO
+
+						struct Vector* sorted_vector = vector_sort(tickets, sortBySeverity);
+						printTicketsArray(sorted_vector);
+						vector_destroy(sorted_vector);
+						break;
+					}
+					case('5'): // Filter by stakeholders
+					{
+						wchar_t* ct;
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: `ctx` not `ct`

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 18]
```diff
@@ -25,11 +25,38 @@ bool filterByStatus(void* ticket, void* status)
...
+}
 bool sortByTitle(void* ticket1, void* ticket2)
 {
 	return wcscmp(ticket_getTitle(ticket1), ticket_getTitle(ticket2)) > 0;
 }
+bool filterByStakeholders(void* ticket1, void*ticket2)
+{
+	return wcscmp(ticket_getstakeholders(ticket1), ticket_getstakeholders(ticket2))!= 0;
+}
+bool searchByCustomerEmail(void* ticket1, void*ticket2)
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Why searchByCustomerEmail gets 2 tickets?
Why not ticket and customer email?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 19]
```diff
@@ -25,11 +25,38 @@ bool filterByStatus(void* ticket, void* status)
...
+{
+	return wcscmp(ticket_getCustomerEmail(ticket1), ticket_getCustomerEmail(ticket2)) != 0;
+
+}
+bool searchBytags(void* ticket1, void*ticket2)
+{
+	return wcscmp(ticket_getTags(ticket1), ticket_getTags(ticket2)) != 0;
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Why `searchBytags` gets 2 tickets?
Why not ticket and tags?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 20]
```diff
@@ -77,30 +104,74 @@ void browseTickets()
...
+						struct Vector* filtered_vector = vector_filter(tickets, filterByStakeholders, stakeholders);
+						printTicketsArray(filtered_vector);
+						vector_destroy(filtered_vector);
 						break;
 					}
-					case('5'): // Search by customer email
+					case('6'): // Search by customer email
 					{
-						// TODO
+						wchar_t* ce;
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: `ctx` not `ce`

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 21]
```diff
@@ -77,30 +104,74 @@ void browseTickets()
...
+						vector_destroy(filtered_vector);
 						break;
 					}
-					case('6'): // Search by tags
+					case('7'): // Search by tags
 					{
-						// TODO
+						wchar_t* cta;
+						wchar_t tags[32];
+						wprintf(L"%s\n", L"please input Customer Email");
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Asking for customer email but actually inputting tags.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 22]
```diff
@@ -129,17 +200,54 @@ void browseTickets()
 	}
 }
 
+void Opentickets()
+{
+		wchar_t customerEmail[256];
+		wchar_t title_support[256];
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Why the `_support` suffix?

File: Source Code/Ticket Management System/Ticket Management System.vcxproj (hunk 2 above)
Comment (Ronshxx) [reply Source Code/Ticket Management System/Ticket Management System.vcxproj]: Still here.

File: Source Code/Ticket Management System/Ticket Management System.vcxproj (hunk 3 above)
Comment (Ronshxx) [reply Source Code/Ticket Management System/Ticket Management System.vcxproj]: Still here.

File: Source Code/Ticket Management System/Ticket Management System.vcxproj (hunk 4 above)
Comment (Ronshxx) [reply Source Code/Ticket Management System/Ticket Management System.vcxproj]: Still here.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 23]
```diff
@@ -25,11 +25,33 @@ bool filterByStatus(void* ticket, void* status)
 {
 	return wcscmp(ticket_getStatus(ticket), status) != 0;
 }
+bool filterByTier(void* ticket1, void* tier)
+{
+	return wcscmp(ticket_getTier(ticket1), ticket_getTier(tier)) != 0;
```
Comment (Ronshxx) [inline Source Code/Ticket Management System/supportgiver_menu.c]: `tier` is not a `ticket`, so you can't call `ticket_getTier` on it.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 24]
```diff
@@ -25,11 +25,33 @@ bool filterByStatus(void* ticket, void* status)
...
+}
+bool filterByType(void* ticket1, void* type)
+{
+	return wcscmp(ticket_getType(ticket1), ticket_getType(type)) != 0;
```
Comment (Ronshxx) [inline Source Code/Ticket Management System/supportgiver_menu.c]: `type` is not a `ticket`, so you can't call `ticket_getType` on it.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 25]
```diff
@@ -25,11 +25,33 @@ bool filterByStatus(void* ticket, void* status)
...
 
+}
 bool sortByTitle(void* ticket1, void* ticket2)
 {
 	return wcscmp(ticket_getTitle(ticket1), ticket_getTitle(ticket2)) > 0;
 }
+bool filterByStakeholders(void* ticket1, void* stakeholders)
+{
+	return wcscmp(ticket_getStakeholders(ticket1), ticket_getStakeholders(stakeholders))!= 0;
```
Comment (Ronshxx) [inline Source Code/Ticket Management System/supportgiver_menu.c]: `stakeholders` is not a ticket, so you can't call `ticket_getStakeholders` on it.

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 18 above)
Comment (Ronshxx) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Still here.

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 19 above)
Comment (Ronshxx) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Still here.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 26]
```diff
@@ -77,30 +99,66 @@ void browseTickets()
...
+						wprintf(L"%s\n", L"please input type");
+						wscanf(L"%s", type);
+						wcstok(type, "\n", &ctx);
+						struct Vector* filtered_vector = vector_filter(tickets, filterByType, type);
+						printTicketsArray(filtered_vector);
+						vector_destroy(filtered_vector);
 						break;
 					}
-					case('4'): // Filter by stakeholders
+					case('5'): // Filter by stakeholders
```
Comment (Ronshxx) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Why was 4 skipped?

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 21 above)
Comment (Ronshxx) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Still an issue.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 27]
```diff
@@ -129,17 +200,54 @@ void browseTickets()
 	}
 }
 
+void Opentickets()
+{
+		wchar_t customerEmail[256];
```
[the remark this answers is not in the document]
Comment (Ronshxx) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Still an issue.

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 22 above)
Comment (Ronshxx) [reply Source Code/Ticket Management System/supportgiver_menu.c]: ...

File: Source Code/Ticket Management System/Ticket Management System.vcxproj (hunk 2 above)
Comment (soryy708) [reply Source Code/Ticket Management System/Ticket Management System.vcxproj]: Still here.

File: Source Code/Ticket Management System/Ticket Management System.vcxproj [hunk 28]
```diff
@@ -22,7 +22,7 @@
     <VCProjectVersion>15.0</VCProjectVersion>
     <ProjectGuid>{BCF3B1AC-70F1-40FD-8EF0-6730B6E2D4E7}</ProjectGuid>
     <RootNamespace>TicketManagementSystem</RootNamespace>
-    <WindowsTargetPlatformVersion>10.0.17763.0</WindowsTargetPlatformVersion>
+    <WindowsTargetPlatformVersion>10.0.16299.0</WindowsTargetPlatformVersion>
```
Comment (soryy708) [inline Source Code/Ticket Management System/Ticket Management System.vcxproj]: Can we please keep it at `10.0.17763.0` to not interfere with other peoples settings?

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 26 above)
Comment (soryy708) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Why was `4` skipped?

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 21 above)
Comment (soryy708) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Still...

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 27 above)
[the remark this answers is not in the document]
Comment (soryy708) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Still...

**Comment by soryy708** [comment]
@Ronshxx @lidorz51 please fix

**Comment by soryy708** [comment]
@lidorz51 Please fix merge conflicts.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 29]
```diff
@@ -24,11 +24,33 @@ bool filterByStatus(void* ticket, void* status)
...
+	return wcscmp(ticket_getType(ticket1), type) != 0;
 
+}
 bool sortByTitle(void* ticket1, void* ticket2)
 {
 	return wcscmp(ticket_getTitle(ticket1), ticket_getTitle(ticket2)) > 0;
 }
+bool filterByStakeholders(void* ticket1, void* stakeholders)
+{
+	return wcscmp(ticket_getStakeholders(ticket1), stakeholders)!= 0;
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Is `stakeholders` a string?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 30]
```diff
@@ -24,11 +24,33 @@ bool filterByStatus(void* ticket, void* status)
...
+}
+bool searchByCustomerEmail(void* ticket1, void* customeremail)
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: `customeremail` does not conform to code style guide.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 31]
```diff
@@ -24,11 +24,33 @@ bool filterByStatus(void* ticket, void* status)
...
+{
+	return wcscmp(ticket_getCustomerEmail(ticket1), customeremail) != 0;
+
+}
+bool searchByTags(void* ticket1, void* tag)
+{
+	return wcscmp(ticket_getTags(ticket1), tag) != 0;
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Is `tags` a string?

File: Source Code/Ticket Management System/supportgiver_menu.c (hunk 21 above)
Comment (soryy708) [reply Source Code/Ticket Management System/supportgiver_menu.c]: Still...

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 32]
```diff
@@ -128,17 +186,66 @@ void browseTickets()
...
+		wscanf(L"%s", description_support);
+		wprintf(L"Please input title\n");
+		wscanf(L"%s", title);
+		wprintf(L"Please input type\n");
+		wscanf(L"%s", type_support);
+		wprintf(L"Please input the option of sevrity{1.Medium,2.High,3.Critical,4.Urgent}:\n");
+		scanf("%d",&severity_num);
+		if (severity_num >= 1 && severity_num <= 4)
+		{
+			severity[severity_num];
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: what do we do with this?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 33]
```diff
@@ -128,17 +186,66 @@ void browseTickets()
...
+		}
+		else
+		{
+			printf("invaild option!!!");
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: very professionally excited print.
Also, will crash on some systems due to not being wide.

Use `wprintf`.

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 34]
```diff
@@ -128,17 +186,66 @@ void browseTickets()
...
+		wprintf(L"===============\n");
+		wprintf(L"Please input description\n");
+		wscanf(L"%s", description_support);
+		wprintf(L"Please input title\n");
+		wscanf(L"%s", title);
+		wprintf(L"Please input type\n");
+		wscanf(L"%s", type_support);
+		wprintf(L"Please input the option of sevrity{1.Medium,2.High,3.Critical,4.Urgent}:\n");
+		scanf("%d",&severity_num);
+		if (severity_num >= 1 && severity_num <= 4)
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: `severity_num` is not enforced. Loop while input is invalid?

File: Source Code/Ticket Management System/supportgiver_menu.c [hunk 35]
```diff
@@ -128,17 +186,66 @@ void browseTickets()
...
+		wscanf(L"%s", customer_email);
+		wprintf(L"Creating a ticket\n");
+		wprintf(L"===============\n");
+		wprintf(L"Please input description\n");
+		wscanf(L"%s", description_support);
+		wprintf(L"Please input title\n");
+		wscanf(L"%s", title);
+		wprintf(L"Please input type\n");
+		wscanf(L"%s", type_support);
+		wprintf(L"Please input the option of sevrity{1.Medium,2.High,3.Critical,4.Urgent}:\n");
```
Comment (soryy708) [inline Source Code/Ticket Management System/supportgiver_menu.c]: Spelling error.

**Comment by soryy708** [comment]
Making good progress. Almost done...

**Pull request merged**